### PR TITLE
Expose FS API to the outside world via Module.FS

### PR DESCRIPTION
**Pull request by regular**
if -s MODULARIZE=1, all of the Emscripten API is hidden in a function. This is great for combining it with different modules (I am using browserify for that). It is critical however to be able to configure the emsciptified Module's behaviour. This can be done by passing a Module object to the exported function.

This is all great.

If I want to prepare something in preRun() for example, I have a problem though: Since the code that prepares the Module object lives outside of the MODULARIZEd function created by emscripten, the FS API is not available to it. The code can however access all of Module's properties, which at that stage are being enriched by the code in various `library_*.js` files. The code in preRun() thus has access to the `FS_*` properties. However, this is just a small subset of the API.

Currently that is:
```
  'FS_createFolder',
  'FS_createPath',
  'FS_createDataFile',
  'FS_createPreloadedFile',
  'FS_createLazyFile',
  'FS_createLink',
  'FS_createDevice',
  'FS_unlink',
```

In my use case, I need `FS.makedev()`, `FS.registerDevice()` and more. I think it's easiest to simply expose the FS object itself as a property of Module.

### CODE REVIEW

**Comment by kripken** [comment]
See `EXPORTED_RUNTIME_METHODS` which was recently added in `src/settings.js`. You can just add `FS` to there, or the specific methods that you want.

Is that good enough?

**Comment by regular** [comment]
Oh, didn't see that. Nice! If I understand the commit messages right, the reason for not exporting the entire `FS` object via `Module` is that this would basically disable dead code elimination for all FS methods. Is that correct?
`EXPORTED_RUNTIME_METHODS` is part of your efforts to reduce code size then? On that topic: I'd really be interested in a `NO_NODE` switch.

Anyway, my problem is solved. Thanks!

**Comment by kripken** [comment]
Yes, that's all correct.

What do you mean by `NO_NODE`? Is there much unnecessary node code?

**Comment by regular** [comment]
It is not the quantity of the code, the problem for me was more what it does.
Let me explain my use case: I try to make a bunch of command line tools work in the browser. Using MODULARIZE=1, I get a nice file with a single function generated by emscripten. I turn that file into a CommonJS module by appending a single line to it:
`module.exports=$EXPORT_NAME`. (A PR for appending that line to the output file when, say, `MODULARIZE=2` is coming up, if you are interested, I think it makes sense and is a quick win)

This CommonJS module I can now use with browserify by `require`ing it from another module (that other module wraps the generated function and adds Node-style stream support for std{in,out,err} (1), not because I would target Node, but because it thus becomes compatible with great client-side modules like [bl](https://www.npmjs.com/package/bl), [multiplex](https://www.npmjs.com/package/multiplex), [wokerstream](https://www.npmjs.com/package/workerstream) and many more.)

To test this, I can simply `require` those modules from my test module. The beauty of this approach is, that those tests work in the browser (when browserified) and also in Node without modification. I am using Node in addition to a test-runner running in the browser because it gives me quicker turn-around times for a first test run.)

... or at least it _would_ work in Node, if the Node-specific code wasn't in the generated function – ironically. My first problem was that browserify's [brfs](https://www.npmjs.com/package/brfs) breaks because it sees `require('fs')` in what it thinks to be code for the browser (the misunderstanding being here is that it is code for the browser _and_ for node in the same module). That was pretty easy to work-around by using brfs on my test module only (it's handy for embedding test fixtures into the test) and telling browserify to not parse but just to embed the emscripten-generated file via `--noparse=<filename>` which is advisable anyway because it speeds up the build process.

The second problem was harder: When the test runner runs in Node (for my convenience), the Node-specific code detected that environment and called `process.exit`, causing the test runner to end with an application-specific exit-code which falsely communicates to the outside world that the test passed/failed and caused parts of the test assertions not to be executed at all (I  *think* I need NO_EXIT_RUNTIME to be zero, but now that I am writing this, I realise I might not have investigated that option enough...). First I solved this by appending `var window={}` to fool the ENVIRONMENT-detection code within the generated file. Later I found out that `NODE_STDOUT_FLUSH_WORKAROUND=0` solved this issue for me too.

Long story short: By being able to simply tell emcc to not emit any Node-specific code at all, I would have saved a lot of time spent to make the test-runner work in Node (ironically, as mentioned before). With [nan](https://github.com/nodejs/nan) and formerly with [gyp](https://github.com/TooTallNate/node-gyp/) we have tools to make bindings between NodeJS and native code anyway and run it at native-speed, so, I imagine, the majority of emcc users would happily disable generating Node-specific code as it often is unnecessary. I could be totally wrong though. 

(1) will be available on npm soon.

**Comment by kripken** [comment]
I see. I would be ok with something like a `FORCE_ENVIRONMENT` option that can be one of the `ENVIRONMENT_IS_*` options, like browser, node, etc. Then we would ifdef out all the others but that one, and make sure that an error is shown if it isn't the right environment.
